### PR TITLE
refactor grad accum

### LINKED ISSUE #545: Broken gradient accumulation
## Description

I ran `python examples/scripts/sentiment_tuning.py --log_with wandb` and compared it with `python examples/scripts/sentiment_tuning.py --log_with wandb --mini_batch_size 1 --gradient_accumulation_steps 128`.

Some gradient accumulation runs get the following errors and crash immediately. Other surviving runs do not learn anything 

image

```
278 │    778 │   a, weights = _ravel_and_check_weights(a, weights)                 │
279 │    779 │                                                                     │
280 │ ❱  780 │   bin_edges, uniform_bins = _get_bin_edges(a, bins, range, weights) │
281 │    781 │                                                                     │
282 │    782 │   # Histogram is an integer or a float array depending on the weigh │
283 │    783 │   if weights is None:                                               │
284 │                                                                              │
285 │ /admin/home/costa/.pyenv/versions/trl/lib/python3.8/site-packages/numpy/lib/ │
286 │ histograms.py:426 in _get_bin_edges                                          │
287 │                                                                              │
288 │    423 │   │   if n_equal_bins < 1:                                          │
289 │    424 │   │   │   raise ValueError('`bins` must be positive, when an intege │
290 │    425 │   │                                                                 │
291 │ ❱  426 │   │   first_edge, last_edge = _get_outer_edges(a, range)            │
292 │    427 │                                                                     │
293 │    428 │   elif np.ndim(bins) == 1:                                          │
294 │    429 │   │   bin_edges = np.asarray(bins)                                  │
295 │                                                                              │
296 │ /admin/home/costa/.pyenv/versions/trl/lib/python3.8/site-packages/numpy/lib/ │
297 │ histograms.py:323 in _get_outer_edges                                        │
298 │                                                                              │
299 │    320 │   else:                                                             │
300 │    321 │   │   first_edge, last_edge = a.min(), a.max()                      │
301 │    322 │   │   if not (np.isfinite(first_edge) and np.isfinite(last_edge)):  │
302 │ ❱  323 │   │   │   raise ValueError(                                         │
303 │    324 │   │   │   │   "autodetected range of [{}, {}] is not finite".format │
304 │    325 │                                                                     │
305 │    326 │   # expand empty range to avoid divide by zero                      │
306 ╰──────────────────────────────────────────────────────────────────────────────╯
307 ValueError: autodetected range of [nan, nan] is not finite
```

This is interesting because we have a test case asserting the **gradient equivalence** of these two hyperparameters... Gonna dig deeper.

https://github.com/lvwerra/trl/blob/d0fe348a0a354bb10e0d728d8e9a4d46d4df0cbc/tests/test_ppo_trainer.py#L1018-L1019
https://github.com/lvwerra/trl/blob/d0fe348a0a354bb10e0d728d8e9a4d46d4df0cbc/tests/test_ppo_trainer.py#L1042-L1043

### PR DESCRIPTION
**Pull request by vwxyzjn**
Closes #545 and #548. This PR does a few things:

* **introduces "micro_batch"**, a more clarified terminology for gradient accumulation. (E.g., see the terminology adoption in [nanoGPT](https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/train.py#L289))
    ```python
    import numpy as np
    batch_size = 8
    mini_batch_size = 4
    gradient_accumulation_steps = 2
    micro_batch_size = mini_batch_size // gradient_accumulation_steps
    data = np.arange(batch_size)
    for epoch in range(2):
        b_inds = np.random.permutation(batch_size)
        print("epoch: ", epoch, "b_inds: ", b_inds)
        for mini_batch_start in range(0, batch_size, mini_batch_size):
            mini_batch_end = mini_batch_start + mini_batch_size
            mini_batch_inds = b_inds[mini_batch_start:mini_batch_end]
            
            # `optimizer.zero_grad()` set optimizer to zero for gradient accumulation
            for micro_batch_start in range(0, mini_batch_size, micro_batch_size):
                micro_batch_end = micro_batch_start + micro_batch_size 
                micro_batch_inds = mini_batch_inds[micro_batch_start:micro_batch_end]
                print("actual training pass on", data[micro_batch_inds])
            # `optimizer.step()`
    ```
    ```
    epoch:  0 b_inds:  [3 7 2 1 4 5 6 0]
    actual training pass on [3 7]
    actual training pass on [2 1]
    actual training pass on [4 5]
    actual training pass on [6 0]
    epoch:  1 b_inds:  [6 5 2 0 7 3 1 4]
    actual training pass on [6 5]
    actual training pass on [2 0]
    actual training pass on [7 3]
    actual training pass on [1 4]
    ``` 
* **Only calculate advantages once**. As pointed out by #548, there is no need to calculate advantages at every forward pass. This PR refactors the learner only to calculate the advantage once.

I then conducted end-to-end experiments, which demonstrated near-identical learning curves when running with `accelerate launch --num_processes 1 examples/scripts/sentiment_tuning.py --log_with wandb --gradient_accumulation_steps 128` and `python examples/scripts/sentiment_tuning.py --log_with wandb` in the main branch.

image

To further confirmed it works across the board, I also tested with `gpt2-xl` and multi-GPU seemed to work perfectly

image

Potential tests to do in the future 
* Models other GPT2 such as starcoder
* DeepSpeed and peft tests 
* test if whitening the global advantages matters in the case of multi-GPU.

### CODE REVIEW

**Comment by vwxyzjn** [comment]
By following accelerate's docs, we should do

```
for batch in training_dataloader:
    with accelerator.accumulate(model):
        inputs, targets = batch
        outputs = model(inputs)
        loss = loss_function(outputs, targets)
        accelerator.backward(loss)
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad()
```

I believe the implementation is correct but for some reason if I do `python examples/scripts/sentiment_tuning.py  --gradient_accumulation_steps 128` it gives me `'ppo/loss/policy': nan, 'ppo/loss/value': nan, 'ppo/loss/total': nan, 'ppo/policy/entropy': nan,` in `stats`... not sure what did I miss.

**Comment by vwxyzjn** [comment]
I fixed all the test case except 
image

File: trl/trainer/ppo_trainer.py
Comment (zfang) [inline trl/trainer/ppo_trainer.py:698]: Hi, 

I'm trying to understand the motivation behind the introduction of `micro_batch` in addition to `mini_batch`.

For example, given the contrived PPO config:
```
mini_batch_size = 4
gradient_accumulation_steps = 2
micro_batch_size = mini_batch_size / gradient_accumulation_steps = 2
```
(A side question: why even make `micro_batch_size` configurable when you will overwrite it with `mini_batch_size / gradient_accumulation_steps`?)

Based on my understanding:
Original:
- Train a batch of 4 (`mini_batch_size`), and zero out gradients for every 2 (`gradient_accumulation_steps`) batches;

New:
- Train a batch of 2 (`micro_batch_size`), and always zero out gradients;
- This is equivalent to setting `mini_batch_size = 2` and `gradient_accumulation_steps = 1` in the original implementation.

So it looks like
1. The behaviors between the original and new implementations are different based on the same `mini_batch_size` and `gradient_accumulation_steps` config values, which could make this change not backward compatible;
2. The additional level of micro batch for loop appears to be unnecessary and further increase code complexity when you can simply use different `mini_batch_size` and `gradient_accumulation_steps` config values to achieve the same behavior;
3. `gradient_accumulation_steps` in the new implementation loses its intended meaning because we do not accumulate gradients at all with micro batches.

Thanks and sorry if I misunderstood anything,

Felix

Comment (vwxyzjn) [reply trl/trainer/ppo_trainer.py:698]: Hi Felix, thanks for the thoughtful reply. 

> A side question: why even make micro_batch_size configurable when you will overwrite it with mini_batch_size / gradient_accumulation_steps?

Because then `micro_batch_size` will be recorded to wandb for experiment tracking to my understanding. 

> Train a batch of 2 (micro_batch_size), and always zero out gradients;... gradient_accumulation_steps in the new implementation loses its intended meaning because we do not accumulate gradients at all with micro batches.

It looks this way but in reality, it's doing something different. Because we hand off gradient accumulation to accelerate, you are supposed to call `optimizer.zero_grad()` even though typically you wouldn't (see https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation#the-finished-code)

> The additional level of micro batch for loop appears to be unnecessary and further increase code complexity when you can simply use different mini_batch_size and gradient_accumulation_steps config values to achieve the same behavior;

I guess this is more of a stylistic thing. We really have three levels of "batch sizes":

1. `batch_size` is the amount of rollout data, e,g., 8 data points
2. `minibatch_size` is the amount of data that you are actually doing a  `zero_grad()`, `loss.backward()`, and `optimizer.step()`, e.g., 4 data points
3. `microbatch_size` if the `minibatch_size` is too large to fit in memory, so we `zero_grad()`, partition the `minibatch` and do multiple `loss.backward()` and then `optimizer.step()`. 

Having the terminology like this makes it clear what the real `minibatch_size` is, so that we do not confound it with `gradient_accumulation_steps`. 

From the user's perspective, one set of hyperparameters should always reliably reproduce similar learning curves. By the new API design in this PR, `batch_size=8, minibatch_size=4` will reliably reproduce similar learning curves regardless of how many `gradient_accumulation_steps` there is, so when doing experiment management / analysis we can just group /filter by `batch_size` and `minibatch_size`. The existing implementation, however, would require us to calculate what's the "real size of the data we perform an `optimizer.step()` on".

File: trl/core.py
Comment (younesbelkada) [inline trl/core.py:129]: Maybe let's add a warning here in case `mask_sum==1`?

File: trl/trainer/ppo_trainer.py
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:675]: Why weren't doing `torch.no_grad()` at the first place? Perhaps we just forgot to add it?

File: trl/trainer/ppo_trainer.py
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:1005]: Nice!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for this refactor ! I think this makes sense, I agree this should make the new version to `0.5.0` for the introduced small breaking change. Let's maybe add 🚨🚨🚨 in the PR title so that it will be clear for developers that rely on TRL that this is a breaking change.
I also agree with: https://github.com/lvwerra/trl/pull/546/files#r1270503430 
I added one open question
Let's wait for @lvwerra 's review before merging
Once this gets merged, let's close: https://github.com/lvwerra/trl/pull/584
Thanks!

File: trl/trainer/ppo_trainer.py
Comment (vwxyzjn) [reply trl/trainer/ppo_trainer.py:675]: Haha I guess we just forgot about it. OAI also did remove the gradient :) (https://github.com/openai/lm-human-preferences/blob/cbfd210bb8b08f6bc5c26878c10984b90f516c66/lm_human_preferences/train_policy.py#L339)

File: trl/core.py
Comment (vwxyzjn) [reply trl/core.py:129]: I think we should just error out actually — it's gonna create nans anyway. Let's wait for @lvwerra's review and address this comment together

File: trl/trainer/ppo_trainer.py
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:675]: I could swear it was there at some point :) thanks for putting it back!

File: trl/core.py
Comment (lvwerra) [reply trl/core.py:129]: Yeah error out sounds good to me here!

**Comment by vwxyzjn** [comment]
The latest commit works out! 

image

File: trl/trainer/ppo_trainer.py
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:706]: I don't think that belongs here anymore.
```suggestion
                # set optimizer to zero for gradient accumulation
```

**Review by lvwerra — APPROVED** [review]
PR looks super clean to me! Thanks a lot for fixing this!

**Review by younesbelkada — APPROVED** [review]
Thanks very much for this great work and investigation @vwxyzjn !

**Pull request merged**
